File: typescript/src/scheduled.ts
import { AnnualGrantService } from "./annual.js";
import type { PlanCatalog } from "./catalog.js";
import type { Database } from "./database.js";
import { ReconciliationService } from "./reconcile.js";
import type {
  ReconciliationGateway,
  ReconciliationProcessor,
} from "./reconcile.js";
import type { ProcessResult, SubscriptionSnapshot } from "./types.js";

const MAX_SERVERLESS_BATCH = 100;

export type CreditPackReconcileOutcome =
  | "reconciled"
  | "idle"
  | "lost_lease"
  | "unavailable"
  | "failed";

export interface CreditPackReconcileSummary {
  readonly outcome: CreditPackReconcileOutcome;
}

export interface ScheduledCreditPackReconciliation {
  reconcileDue(options: {
    readonly limit: number;
  }): Promise<readonly CreditPackReconcileSummary[]>;
}

export interface ScheduledGateway extends ReconciliationGateway {
  subscriptionSnapshot(subscriptionId: string): Promise<SubscriptionSnapshot>;
}

export interface ScheduledKernel {
  readonly database: Database;
  readonly catalog: PlanCatalog;
  readonly gateway: ScheduledGateway;
  requireServices(): {
    readonly processor: ReconciliationProcessor & {
      readonly productLine: string;
    };
    readonly creditPackReconciliation: ScheduledCreditPackReconciliation;
  };
}

export interface AnnualGrantWorker {
  dueAccounts(
    now?: undefined,
    options?: { readonly limit?: number },
  ): Promise<
    readonly { readonly id: string; readonly stripe_subscription_id: string }[]
  >;
  recordFailure(
    accountId: string,
    subscriptionId: string,
    reason: string,
  ): Promise<void>;
  deferCandidate(accountId: string): Promise<void>;
  grantDue(
    accountId: string,
    now: undefined,
    snapshot: SubscriptionSnapshot,
  ): Promise<ProcessResult>;
}

export interface ReconciliationWorker {
  databaseNow(): Promise<string>;
  candidates(
    now?: undefined,
    options?: {
      readonly limit?: number;
      readonly attemptedBefore?: string;
    },
  ): Promise<readonly { readonly id: string }[]>;
  reconcileAccount(accountId: string): Promise<ProcessResult>;
}

export interface ScheduledWorkerFactories {
  annual(kernel: ScheduledKernel): AnnualGrantWorker;
  reconciliation(kernel: ScheduledKernel): ReconciliationWorker;
}

const DEFAULT_FACTORIES: ScheduledWorkerFactories = {
  annual(kernel) {
    const services = kernel.requireServices();
    return new AnnualGrantService(
      kernel.database,
      kernel.catalog,
      services.processor,
    );
  },
  reconciliation(kernel) {
    const services = kernel.requireServices();
    return new ReconciliationService(
      kernel.database,
      services.processor,
      kernel.gateway,
    );
  },
};

function boundedBatchLimit(value: number, field: string): number {
  if (
    !Number.isSafeInteger(value) ||
    typeof value !== "number" ||
    value < 1 ||
    value > MAX_SERVERLESS_BATCH
  ) {
    throw new RangeError(
      `${field} must be an integer between 1 and ${MAX_SERVERLESS_BATCH}`,
    );
  }
  return value;
}

export class AnnualGrantBatchResult {
  public readonly attempted: number;
  public readonly handled: number;
  public readonly replayed: number;
  public readonly ignored: number;
  public readonly failures: number;

  public constructor(input: {
    readonly attempted: number;
    readonly handled: number;
    readonly replayed: number;
    readonly ignored: number;
    readonly failures: number;
  }) {
    Object.assign(this, input);
    this.attempted = input.attempted;
    this.handled = input.handled;
    this.replayed = input.replayed;
    this.ignored = input.ignored;
    this.failures = input.failures;
  }

  public get ok(): boolean {
    return this.failures === 0;
  }

  public publicSummary(): Readonly<Record<string, boolean | number>> {
    return {
      ok: this.ok,
      attempted: this.attempted,
      handled: this.handled,
      replayed: this.replayed,
      ignored: this.ignored,
      failures: this.failures,
    };
  }
}

export class ReconciliationBatchResult {
  public readonly accountsAttempted: number;
  public readonly accountsHandled: number;
  public readonly accountsReplayed: number;
  public readonly accountsIgnored: number;
  public readonly packsAttempted: number;
  public readonly packsReconciled: number;
  public readonly packsIdle: number;
  public readonly packsDeferred: number;
  public readonly failures: number;

  public constructor(input: {
    readonly accountsAttempted: number;
    readonly accountsHandled: number;
    readonly accountsReplayed: number;
    readonly accountsIgnored: number;
    readonly packsAttempted: number;
    readonly packsReconciled: number;
    readonly packsIdle: number;
    readonly packsDeferred: number;
    readonly failures: number;
  }) {
    this.accountsAttempted = input.accountsAttempted;
    this.accountsHandled = input.accountsHandled;
    this.accountsReplayed = input.accountsReplayed;
    this.accountsIgnored = input.accountsIgnored;
    this.packsAttempted = input.packsAttempted;
    this.packsReconciled = input.packsReconciled;
    this.packsIdle = input.packsIdle;
    this.packsDeferred = input.packsDeferred;
    this.failures = input.failures;
  }

  public get ok(): boolean {
    return this.failures === 0;
  }

  public publicSummary(): Readonly<Record<string, boolean | number>> {
    return {
      ok: this.ok,
      accounts_attempted: this.accountsAttempted,
      accounts_handled: this.accountsHandled,
      accounts_replayed: this.accountsReplayed,
      accounts_ignored: this.accountsIgnored,
      packs_attempted: this.packsAttempted,
      packs_reconciled: this.packsReconciled,
      packs_idle: this.packsIdle,
      packs_deferred: this.packsDeferred,
      failures: this.failures,
    };
  }
}

export async function runAnnualGrantBatch(
  kernel: ScheduledKernel,
  options: {
    readonly limit?: number;
    readonly factories?: ScheduledWorkerFactories;
  } = {},
): Promise<AnnualGrantBatchResult> {
  const limit = boundedBatchLimit(
    options.limit ?? 25,
    "annual grant batch limit",
  );
  const worker = (options.factories ?? DEFAULT_FACTORIES).annual(kernel);
  const candidates = await worker.dueAccounts(undefined, { limit });
  let handled = 0;
  let replayed = 0;
  let ignored = 0;
  let failures = 0;
  for (const candidate of candidates) {
    let snapshot: SubscriptionSnapshot;
    try {
      snapshot = await kernel.gateway.subscriptionSnapshot(
        candidate.stripe_subscription_id,
      );
    } catch (error) {
      failures += 1;
      try {
        await worker.recordFailure(
          candidate.id,
          candidate.stripe_subscription_id,
          `subscription snapshot failed: ${
            error instanceof Error ? error.constructor.name : "UnknownError"
          }`,
        );
      } catch {
        // The non-zero summary still asks the scheduler to retry.
      }
      try {
        await worker.deferCandidate(candidate.id);
      } catch {
        // The candidate remains visible for the next bounded pass.
      }
      continue;
    }
    let result: ProcessResult;
    try {
      result = await worker.grantDue(candidate.id, undefined, snapshot);
    } catch {
      failures += 1;
      continue;
    }
    if (result.outcome === "handled") {
      handled += 1;
    } else if (result.outcome === "replayed") {
      replayed += 1;
    } else if (result.outcome === "ignored") {
      ignored += 1;
      try {
        await worker.deferCandidate(candidate.id);
      } catch {
        failures += 1;
      }
    } else {
      failures += 1;
    }
  }
  return new AnnualGrantBatchResult({
    attempted: candidates.length,
    handled,
    replayed,
    ignored,
    failures,
  });
}

export async function runReconciliationBatch(
  kernel: ScheduledKernel,
  options: {
    readonly accountLimit?: number;
    readonly creditPackLimit?: number;
    readonly factories?: ScheduledWorkerFactories;
  } = {},
): Promise<ReconciliationBatchResult> {
  const accountLimit = boundedBatchLimit(
    options.accountLimit ?? 20,
    "reconciliation account batch limit",
  );
  const creditPackLimit = boundedBatchLimit(
    options.creditPackLimit ?? 20,
    "credit-pack reconciliation batch limit",
  );
  const services = kernel.requireServices();
  const worker = (options.factories ?? DEFAULT_FACTORIES).reconciliation(
    kernel,
  );
  const runStarted = await worker.databaseNow();
  const candidates = await worker.candidates(undefined, {
    limit: accountLimit,
    attemptedBefore: runStarted,
  });
  let accountsHandled = 0;
  let accountsReplayed = 0;
  let accountsIgnored = 0;
  let failures = 0;
  for (const candidate of candidates) {
    let result: ProcessResult;
    try {
      result = await worker.reconcileAccount(candidate.id);
    } catch {
      failures += 1;
      continue;
    }
    if (result.outcome === "handled") {
      accountsHandled += 1;
    } else if (result.outcome === "replayed") {
      accountsReplayed += 1;
    } else if (result.outcome === "ignored") {
      accountsIgnored += 1;
    } else {
      failures += 1;
    }
  }
  const packResults = await services.creditPackReconciliation.reconcileDue({
    limit: creditPackLimit,
  });
  const packsReconciled = packResults.filter(
    (result) => result.outcome === "reconciled",
  ).length;
  const packsIdle = packResults.filter(
    (result) => result.outcome === "idle",
  ).length;
  const packsDeferred = packResults.filter(
    (result) =>
      result.outcome === "lost_lease" || result.outcome === "unavailable",
  ).length;
  const packFailures = packResults.filter(
    (result) => result.outcome === "failed",
  ).length;
  const known = packsReconciled + packsIdle + packsDeferred + packFailures;
  failures += packFailures + (packResults.length - known);
  return new ReconciliationBatchResult({
    accountsAttempted: candidates.length,
    accountsHandled,
    accountsReplayed,
    accountsIgnored,
    packsAttempted: packResults.length,
    packsReconciled,
    packsIdle,
    packsDeferred,
    failures,
  });
}

File: src/stripe_entitlements/scheduled.py
from __future__ import annotations

from dataclasses import dataclass

from .annual import AnnualGrantService
from .kernel import BillingKernel
from .reconcile import ReconciliationService

_MAX_SERVERLESS_BATCH = 100


def _bounded_batch_limit(value: int, *, field: str) -> int:
    if type(value) is not int or value < 1 or value > _MAX_SERVERLESS_BATCH:
        raise ValueError(f"{field} must be an integer between 1 and {_MAX_SERVERLESS_BATCH}")
    return value


@dataclass(frozen=True, slots=True)
class AnnualGrantBatchResult:
    """Identity-free summary for one bounded annual-grant scheduler invocation."""

    attempted: int
    handled: int
    replayed: int
    ignored: int
    failures: int

    @property
    def ok(self) -> bool:
        return self.failures == 0

    def public_summary(self) -> dict[str, bool | int]:
        return {
            "ok": self.ok,
            "attempted": self.attempted,
            "handled": self.handled,
            "replayed": self.replayed,
            "ignored": self.ignored,
            "failures": self.failures,
        }


@dataclass(frozen=True, slots=True)
class ReconciliationBatchResult:
    """Identity-free summary for one bounded subscription and pack repair pass."""

    accounts_attempted: int
    accounts_handled: int
    accounts_replayed: int
    accounts_ignored: int
    packs_attempted: int
    packs_reconciled: int
    packs_idle: int
    packs_deferred: int
    failures: int

    @property
    def ok(self) -> bool:
        return self.failures == 0

    def public_summary(self) -> dict[str, bool | int]:
        return {
            "ok": self.ok,
            "accounts_attempted": self.accounts_attempted,
            "accounts_handled": self.accounts_handled,
            "accounts_replayed": self.accounts_replayed,
            "accounts_ignored": self.accounts_ignored,
            "packs_attempted": self.packs_attempted,
            "packs_reconciled": self.packs_reconciled,
            "packs_idle": self.packs_idle,
            "packs_deferred": self.packs_deferred,
            "failures": self.failures,
        }


async def run_annual_grant_batch(
    kernel: BillingKernel,
    *,
    limit: int = 25,
) -> AnnualGrantBatchResult:
    """Run at most ``limit`` annual candidates using the active kernel graph.

    The database row lock and invoice/slot uniqueness guard remain the concurrency
    authority. This wrapper deliberately processes only one page so it fits a bounded
    function invocation; a later scheduler tick safely continues the scan.
    """

    limit = _bounded_batch_limit(limit, field="annual grant batch limit")
    services = kernel.require_services()
    worker = AnnualGrantService(
        kernel.database.require_pool(),
        kernel.catalog,
        services.processor,
    )
    candidates = await worker.due_accounts(None, limit=limit)
    handled = replayed = ignored = failures = 0
    for candidate in candidates:
        account_id = str(candidate["id"])
        subscription_id = str(candidate["stripe_subscription_id"])
        try:
            snapshot = await kernel.gateway.subscription_snapshot(subscription_id)
        except Exception as exc:
            failures += 1
            try:
                await worker.record_failure(
                    account_id,
                    subscription_id,
                    f"subscription snapshot failed: {type(exc).__name__}",
                )
            except Exception:
                # A failed incident write must not stop later candidates. The non-zero
                # summary makes the scheduler retry this idempotent batch.
                pass
            try:
                await worker.defer_candidate(account_id)
            except Exception:
                pass
            continue
        try:
            result = await worker.grant_due(account_id, None, snapshot)
        except Exception:
            failures += 1
            continue
        if result.outcome == "handled":
            handled += 1
        elif result.outcome == "replayed":
            replayed += 1
        elif result.outcome == "ignored":
            ignored += 1
            try:
                await worker.defer_candidate(account_id)
            except Exception:
                failures += 1
        else:
            failures += 1
    return AnnualGrantBatchResult(
        attempted=len(candidates),
        handled=handled,
        replayed=replayed,
        ignored=ignored,
        failures=failures,
    )


async def run_reconciliation_batch(
    kernel: BillingKernel,
    *,
    account_limit: int = 20,
    credit_pack_limit: int = 20,
) -> ReconciliationBatchResult:
    """Run one bounded account-repair page and one fenced credit-pack page."""

    account_limit = _bounded_batch_limit(account_limit, field="reconciliation account batch limit")
    credit_pack_limit = _bounded_batch_limit(
        credit_pack_limit, field="credit-pack reconciliation batch limit"
    )
    services = kernel.require_services()
    worker = ReconciliationService(
        kernel.database.require_pool(),
        services.processor,
        kernel.gateway,
    )
    run_started = await worker.database_now()
    candidates = await worker.candidates(
        None,
        limit=account_limit,
        attempted_before=run_started,
    )
    handled = replayed = ignored = failures = 0
    for candidate in candidates:
        try:
            result = await worker.reconcile_account(str(candidate["id"]))
        except Exception:
            failures += 1
            continue
        if result.outcome == "handled":
            handled += 1
        elif result.outcome == "replayed":
            replayed += 1
        elif result.outcome == "ignored":
            ignored += 1
        else:
            failures += 1

    pack_results = await services.credit_pack_reconciliation.reconcile_due(limit=credit_pack_limit)
    packs_reconciled = sum(result.outcome == "reconciled" for result in pack_results)
    packs_idle = sum(result.outcome == "idle" for result in pack_results)
    packs_deferred = sum(result.outcome in {"lost_lease", "unavailable"} for result in pack_results)
    failures += sum(result.outcome == "failed" for result in pack_results)
    known_pack_outcomes = (
        packs_reconciled
        + packs_idle
        + packs_deferred
        + sum(result.outcome == "failed" for result in pack_results)
    )
    failures += len(pack_results) - known_pack_outcomes
    return ReconciliationBatchResult(
        accounts_attempted=len(candidates),
        accounts_handled=handled,
        accounts_replayed=replayed,
        accounts_ignored=ignored,
        packs_attempted=len(pack_results),
        packs_reconciled=packs_reconciled,
        packs_idle=packs_idle,
        packs_deferred=packs_deferred,
        failures=failures,
    )


__all__ = [
    "AnnualGrantBatchResult",
    "ReconciliationBatchResult",
    "run_annual_grant_batch",
    "run_reconciliation_batch",
]
